src/ExadelMentorship.Jobs/ReportJob.cs

```diff
@@ -0,0 +1,43 @@
+﻿using ExadelMentorship.BusinessLogic.Interfaces;
+using ExadelMentorship.BusinessLogic.Interfaces.MessageBus;
+using ExadelMentorship.DataAccess;
+using ExadelMentorship.DataAccess.Entities;
+using Hangfire;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace ExadelMentorship.Jobs
+{
+    public class ReportJob : BackgroundService
+    {
+        private readonly IMessageProducer _messagePublisher;
+        private IServiceProvider _services;
+
+        public ReportJob(IMessageProducer messagePublisher,IServiceProvider services)
+        {
+            _messagePublisher = messagePublisher;
+            _services = services;
+        }
+        protected async override Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            using var scope = _services.CreateScope();
+            var all = await scope.ServiceProvider.GetRequiredService<IReportUserRepo>().GetAll();
+            foreach (var item in all)
+            {
+                RecurringJob.AddOrUpdate(
+                    $"ReportSendingJobFor_{item.UserName}",
+                    () => this.Execute(item),
+                    $"0 * {item.Period} * * ?"
+                );
+            }
+        }
+        public async Task Execute(ReportUser reportUser)
+        {
+            using var scope = _services.CreateScope();
```

Comment:
1. Why need to create a DI scope? "Because `IReportServices` is registered as scoped" is not an answer.
2. Why not just inject it in the constructor of ReportJob class?

Reply:
because HostedService is singleton
"InvalidOperationException: Cannot consume scoped service 'ExadelMentorship.BusinessLogic.Interfaces.IReportServices' from singleton 'Microsoft.Extensions.Hosting.IHostedService'.
"

Reply:
It's a chain, IReportService  depends on IReportUserRepo, which depends on ApplicationDbContext(which one can not be singleton, it is socped), that forces IReportUserRepo to be scoped, which forces IReportService   to be scoped

Reply:
IReportService becoming scoped because DbContext is scoped is not good decision. Make IReportService singleton, and resolve DbContext there as scoped, as discussed in the call.

Reply:
done